agents/ahv/fence_ahv.py

```diff
@@ -0,0 +1,744 @@
+"""
+AHV Fence agent
+
+Compatible with AHV v4 API
+"""
+
+
+import atexit
+import logging
+import sys
+import time
+import uuid
+import requests
+
+sys.path.append("@FENCEAGENTSLIBDIR@")
+import fencing
+
+
+V4_VERSION = '4.0'
+MIN_TIMEOUT = 60
+PC_PORT = 9440
+
+
+class NutanixClientException(Exception):
+        """
+        Exception class for Nutanix NutanixClient
+        """
+
+class AHVFenceAgentException(Exception):
+        """
+        Exception class for AHV Fence Agent
+        """
+
+
+class NutanixClient:
+        """
+        Nutanix REST Client
+        """
+        def __init__(self, username, password):
+                self.username = username
+                self.password = password
+                self.valid_status_codes = [200, 202]
+
+        def request(self, url, method='GET', headers=None, **kwargs):
+                """
+                Process requests based on the requested method.
+
+                Args:
+                url(str): URL of the API end point.
+                method(str): HTTP request method.
+                headers(dict): HTTP request headers
+                **kwargs: Any other keyword args relevant to the method.
+
+                Returns:
+                response: response object.
+                """
+                session = requests.Session()
```

Comment:
do we want to create a session on every request?

Reply:
If I move the session into __init__() then I keep getting rate limit error.

requests.exceptions.HTTPError: 429 Client Error: TOO MANY REQUESTS for url: https://10.101.63.173:9440/api/vmm/v4.0/ahv/config/vms?$filter=name%20eq%20'TestVM1'

Am I missing something? Is there a way to avoid this error and preserve the session?

Reply:
Ah I see, this is probably referring to the rate limit. This means that server is limiting the number of requests you can make over a specified time period. I guess we can add retry to that logic but for now this should be ok as long as the stress testing won't cause any issue